onnxscript/rewriter/ort_fusions/sdpa.py

```diff
@@ -41,173 +25,125 @@ def pattern(
-                return check_result.fail(
-                    "Query scale is not a scalar.",
-                    query_scale,
-                )
-            # Ensure the scaling factor for key is the same as for query
-            if (key_scale_value := _ir_utils.get_singleton_value(key_scale)) is None:
-                return check_result.fail(
-                    "Key scale is not a scalar.",
-                    key_scale,
-                )
-            if not math.isclose(query_scale_value, key_scale_value, rel_tol=1e-3):
-                return check_result.fail(
-                    "Query and key scales are not equal.",
-                    query_scale,
-                )
-            if not math.isclose(query_scale_value, sqrt_scaling_factor, rel_tol=1e-3):
-                self._scale = query_scale_value * query_scale_value
-            else:
-                # Pass no scaling factor to SDPA, SDPA will use the default scaling factor
-                self._scale = None
-        else:
-            # Check if qk_scale is a scalar == expected_scaling_factor)
-            # If it is a scalar but != sqrt(expected_scaling_factor), a custom scale is being used
-            if (qk_scale_value := _ir_utils.get_singleton_value(qk_scale)) is None:
-                return check_result.fail(
-                    "QK scale is not a scalar.",
-                    qk_scale,
-                )
-            if not math.isclose(qk_scale_value, expected_scaling_factor, rel_tol=1e-3):
-                self._scale = qk_scale_value
+
+        bindings: dict[str, Dim] = {}
+
+        # Check that query/key/value have the expected shapes:
+        # They all should have same batch-size (B) and number of heads (H). Conceptually, it is
+        # different for Q and K/V, but the certain op implementations require them to be the same,
+        # which is usually achieved via tiling/expanding K/V num-heads to match Q num-heads.
+        # Query and Key should have same head-size (Dh) while value can have different head-size (Dv).
+        # Key and Value should have same sequence length (Skv), while Query can have different sequence length (S).
+        _fusion_utils.check_shape(bindings, query, ["B", "H", "S", "Dh"])
+        _fusion_utils.check_shape(bindings, key_transposed, ["B", "H", "Dh", "Skv"])
+        _fusion_utils.check_shape(bindings, value, ["B", "H", "Skv", "Dv"])
+
+        def get_scale_value(tag_name: str, scale_name: str) -> float:
+            scaling_type = match_bindings.get(tag_name, "None")
+            if scaling_type == "None":
+                return 1.0
             else:
-                # Pass no scaling factor to SDPA, SDPA will use the default scaling factor
-                self._scale = None
+                scale = match_bindings.get(scale_name)
+                value = _ir_utils.get_singleton_value(scale)
+                if value is None:
+                    raise MatchFailureError(f"{scale_name} is not a scalar.", scale)
```

Comment:
Curious what the benefits are with raising the error over check_result.fail?

Reply:
Compositionality. If we return an error-value here, then the caller has to check for that and return that. An exception avoids the need for that. Same thing with utility functions we can write that are called from various fusion rules.